// homebridge-synology-surveillance-homemode-dsm7/index.js
const axios = require('axios');

let Service, Characteristic;

module.exports = (homebridge) => {
    Service = homebridge.hap.Service;
    Characteristic = homebridge.hap.Characteristic;
    homebridge.registerAccessory(
        'homebridge-synology-surveillance-homemode-dsm7',
        'SynologyHomeMode',
        SynologyHomeModeAccessory
    );
};

class SynologyHomeModeAccessory {

    constructor(log, config) {
        this.log = log;
        this.host = config.host;
        this.username = config.username;
        this.password = config.password;
        this.port = config.port || 5001;
        this.protocol = 'https';
        this.sid = null;

        this.service = new Service.Switch(config.name);

        this.service
            .getCharacteristic(Characteristic.On)
            .onSet(this.setHomeMode.bind(this))
            .onGet(this.getHomeMode.bind(this));
    }

    getServices() {
        return [this.service];
    }

    async login() {
        const url = `${this.protocol}://${this.host}:${this.port}/webapi/auth.cgi`;

        const response = await axios.get(url, {
            params: {
                api: 'SYNO.API.Auth',
                method: 'login',
                version: 7,
                account: this.username,
                passwd: this.password,
                session: 'SurveillanceStation',
                format: 'sid'
            },
            httpsAgent: new (require('https').Agent)({
                rejectUnauthorized: false
            })
        });

        if (!response.data.success) {
            throw new Error('Authentication failed');
        }

        this.sid = response.data.data.sid;
    }

    async setHomeMode(value) {
        if (!this.sid) await this.login();

        const url = `${this.protocol}://${this.host}:${this.port}/webapi/entry.cgi`;

        await axios.get(url, {
            params: {
                api: 'SYNO.SurveillanceStation.HomeMode',
                method: 'Switch',
                version: 1,
                on: value,
                _sid: this.sid
            },
            httpsAgent: new (require('https').Agent)({
                rejectUnauthorized: false
            })
        });

        this.log(`HomeMode set to: ${value}`);
    }

    async getHomeMode() {
        if (!this.sid) await this.login();

        const url = `${this.protocol}://${this.host}:${this.port}/webapi/entry.cgi`;

        const response = await axios.get(url, {
            params: {
                api: 'SYNO.SurveillanceStation.HomeMode',
                method: 'GetInfo',
                version: 1,
                _sid: this.sid
            },
            httpsAgent: new (require('https').Agent)({
                rejectUnauthorized: false
            })
        });

        return response.data.data.on;
    }
}
